org.eclipse.jgit/src/org/eclipse/jgit/internal/storage/dfs/DfsBlockCache.java

```diff
@@ -1,568 +1,56 @@
-				if (table.compareAndSet(slot, e2, n))
-					break;
-				e2 = table.get(slot);
-			}
-			addToClock(ref, blockSize - v.size());
-		} finally {
-			regionLock.unlock();
-		}
-
-		// If the block size changed from the default, it is possible the block
-		// that was loaded is the wrong block for the requested position.
-		if (v.contains(pack.key, requestedPosition))
-			return v;
-		return getOrLoad(pack, requestedPosition, ctx);
-	}
-
-	@SuppressWarnings("unchecked")
-	private void reserveSpace(int reserve) {
-		clockLock.lock();
-		try {
-			long live = liveBytes + reserve;
-			if (maxBytes < live) {
-				Ref prev = clockHand;
-				Ref hand = clockHand.next;
-				do {
-					if (hand.hot) {
-						// Value was recently touched. Clear
-						// hot and give it another chance.
-						hand.hot = false;
-						prev = hand;
-						hand = hand.next;
-						continue;
-					} else if (prev == hand)
-						break;
-
-					// No recent access since last scan, kill
-					// value and remove from clock.
-					Ref dead = hand;
-					hand = hand.next;
-					prev.next = hand;
-					dead.next = null;
-					dead.value = null;
-					live -= dead.size;
-					dead.pack.cachedSize.addAndGet(-dead.size);
-					statEvict++;
-				} while (maxBytes < live);
-				clockHand = prev;
-			}
-			liveBytes = live;
-		} finally {
-			clockLock.unlock();
-		}
-	}
-
-	private void creditSpace(int credit) {
-		clockLock.lock();
-		liveBytes -= credit;
-		clockLock.unlock();
-	}
-
-	@SuppressWarnings("unchecked")
-	private void addToClock(Ref ref, int credit) {
-		clockLock.lock();
-		try {
-			if (credit != 0)
-				liveBytes -= credit;
-			Ref ptr = clockHand;
-			ref.next = ptr.next;
-			ptr.next = ref;
-			clockHand = ref;
-		} finally {
-			clockLock.unlock();
-		}
-	}
+import java.io.IOException;
 
-	void put(DfsBlock v) {
-		put(v.pack, v.start, v.size(), v);
-	}
+public abstract class DfsBlockCache {
 
-	<T> Ref<T> put(DfsPackKey key, long pos, int size, T v) {
-		int slot = slot(key, pos);
-		HashEntry e1 = table.get(slot);
-		Ref<T> ref = scanRef(e1, key, pos);
-		if (ref != null)
-			return ref;
+    private static volatile DfsBlockCache cache;
 
-		reserveSpace(size);
-		ReentrantLock regionLock = lockFor(key, pos);
-		regionLock.lock();
-		try {
-			HashEntry e2 = table.get(slot);
-			if (e2 != e1) {
-				ref = scanRef(e2, key, pos);
-				if (ref != null) {
-					creditSpace(size);
-					return ref;
-				}
-			}
+    static {
+        DefaultDfsBlockCache.reconfigure(new DefaultDfsBlockCacheConfig());
+    }
 
-			key.cachedSize.addAndGet(size);
-			ref = new Ref<T>(key, pos, size, v);
-			ref.hot = true;
-			for (;;) {
-				HashEntry n = new HashEntry(clean(e2), ref);
-				if (table.compareAndSet(slot, e2, n))
-					break;
-				e2 = table.get(slot);
-			}
-			addToClock(ref, 0);
-		} finally {
-			regionLock.unlock();
-		}
-		return ref;
-	}
+    public static void resetCache(DfsBlockCache newCache) {
+        DfsBlockCache oldCache = cache;
+        cache = newCache;
```

Comment:
probably want to make the `cache` an `AtomicReference` to do `DfsBlockCache oldCache = cache.getAndSet(newCache);` as there's currently a possible race condition here for multiple concurrent calls.

nit: might want to name method as `set` or `setInstance`